pkg/kubelet/secret_manager.go

```diff
@@ -39,3 +54,181 @@ func newSimpleSecretManager(kubeClient clientset.Interface) (secretManager, erro
 func (s *simpleSecretManager) GetSecret(namespace, name string) (*api.Secret, error) {
 	return s.kubeClient.Core().Secrets(namespace).Get(name)
 }
+
+func (s *simpleSecretManager) RegisterPod(pod *api.Pod) {
+}
+
+func (s *simpleSecretManager) UnregisterPod(pod *api.Pod) {
+}
+
+// secretStoreItems is a single item stored in secretStore.
+type secretStoreItem struct {
+	secret   *api.Secret
+	refCount int
+}
+
+// secretStore is a local cache of secrets.
+type secretStore struct {
+	kubeClient clientset.Interface
+
+	sync.Mutex
+	items map[string]*secretStoreItem
+}
+
+func newSecretStore(kubeClient clientset.Interface) *secretStore {
+	ss := &secretStore{
+		kubeClient: kubeClient,
+		items:      make(map[string]*secretStoreItem),
+	}
+	go wait.NonSlidingUntil(func() { ss.Refresh() }, time.Minute, wait.NeverStop)
+	return ss
+}
+
+func keyFunc(namespace, name string) string {
+	return namespace + "/" + name
+}
+
+func namespaceNameFunc(key string) (string, string, error) {
+	items := strings.Split(key, "/")
+	if len(items) != 2 {
+		return "", "", fmt.Errorf("invalid key: %s", key)
+	}
+	return items[0], items[1], nil
+}
+
+func (s *secretStore) Add(namespace, name string) {
+	key := keyFunc(namespace, name)
+	secret, err := s.kubeClient.Core().Secrets(namespace).Get(name)
+	if err != nil {
+		glog.Errorf("Unable to retrieve a secret %s/%s: %v", namespace, name, err)
+	}
+
+	s.Lock()
+	defer s.Unlock()
+	if item, ok := s.items[key]; ok {
+		item.secret = secret
+		item.refCount++
+	} else {
+		s.items[key] = &secretStoreItem{secret: secret, refCount: 1}
+	}
+}
+
+func (s *secretStore) Delete(namespace, name string) {
+	key := keyFunc(namespace, name)
+
+	s.Lock()
+	defer s.Unlock()
+	if item, ok := s.items[key]; ok {
+		item.refCount--
+		if item.refCount == 0 {
+			delete(s.items, key)
+		}
+	}
+}
+
+func (s *secretStore) Get(namespace, name string) (*api.Secret, error) {
+	key := keyFunc(namespace, name)
+
+	s.Lock()
+	defer s.Unlock()
+	if item, ok := s.items[key]; ok {
+		return item.secret, nil
+	}
+	return nil, fmt.Errorf("secret no present")
+}
+
+func (s *secretStore) Refresh() {
+	s.Lock()
+	keys := make([]string, 0, len(s.items))
+	for key := range s.items {
+		keys = append(keys, key)
+	}
+	s.Unlock()
+
+	secrets := make([]*api.Secret, 0, len(keys))
+	for _, key := range keys {
+		namespace, name, err := namespaceNameFunc(key)
+		if err != nil {
+			glog.Errorf("Unexpected key transform error: %v", err)
+			secrets = append(secrets, nil)
+			continue
+		}
+		secret, err := s.kubeClient.Core().Secrets(namespace).Get(name)
+		if err != nil {
+			glog.Errorf("Unable to retrieve a secret %s/%s: %v", namespace, name, err)
+			secret = nil
+		}
+		secrets = append(secrets, secret)
+	}
+
+	s.Lock()
+	defer s.Unlock()
+	for i, key := range keys {
+		if secrets[i] == nil {
+			// Ignore updating to nil.
+		}
+		if item, ok := s.items[key]; ok {
```

Comment:
Yes, there could be a race. But it will self-repair itself during Refresh. Do you think it's not enough? Should we check resourceVersions here to guarantee that we not change to older version?